File: services/text-service/controllers/health_controller.py
from fastapi import APIRouter, status, HTTPException
from pydantic import BaseModel
from typing import Dict, Optional, Any
import time
import boto3
from datetime import datetime
import os

router = APIRouter(prefix="/health", tags=["health"])


class HealthCheck(BaseModel):
    status: str
    service: str
    timestamp: str
    checks: Dict[str, Dict[str, Any]]
    version: str


@router.get("/", response_model=HealthCheck, status_code=status.HTTP_200_OK)
async def health_check():
    """
    Comprehensive health check with all dependencies
    """
    checks = {}
    overall_status = "healthy"

    try:
        # Check DynamoDB connectivity
        dynamodb_start = time.time()
        try:
            dynamodb = boto3.client(
                "dynamodb", region_name=os.getenv("AWS_REGION", "ap-southeast-1")
            )
            table_name = os.getenv("DYNAMODB_TABLE_NAME", "typerush-dev-texts")
            dynamodb.describe_table(TableName=table_name)
            checks["dynamodb"] = {
                "status": "healthy",
                "latency": round((time.time() - dynamodb_start) * 1000, 2),
            }
        except Exception as e:
            checks["dynamodb"] = {"status": "unhealthy", "error": str(e)}
            overall_status = "unhealthy"

        # Check Bedrock connectivity
        bedrock_start = time.time()
        try:
            bedrock = boto3.client(
                "bedrock-runtime", region_name=os.getenv("AWS_REGION", "ap-southeast-1")
            )
            # Just verify client creation (no actual invocation to avoid costs)
            checks["bedrock"] = {
                "status": "healthy",
                "latency": round((time.time() - bedrock_start) * 1000, 2),
            }
        except Exception as e:
            checks["bedrock"] = {"status": "unhealthy", "error": str(e)}
            overall_status = "unhealthy"

        health_response = HealthCheck(
            status=overall_status,
            service="text-service",
            timestamp=datetime.utcnow().isoformat(),
            checks=checks,
            version=os.getenv("APP_VERSION", "1.0.0"),
        )

        if overall_status == "unhealthy":
            raise HTTPException(status_code=503, detail=health_response.dict())

        return health_response

    except HTTPException:
        raise
    except Exception as e:
        raise HTTPException(
            status_code=503,
            detail={
                "status": "unhealthy",
                "service": "text-service",
                "timestamp": datetime.utcnow().isoformat(),
                "checks": checks,
                "version": os.getenv("APP_VERSION", "1.0.0"),
                "error": str(e),
            },
        )


@router.get("/live", status_code=status.HTTP_200_OK)
async def liveness_check():
    """
    Simple liveness probe - no dependencies checked
    """
    return {
        "status": "alive",
        "service": "text-service",
        "timestamp": datetime.utcnow().isoformat(),
    }


@router.get("/ready", status_code=status.HTTP_200_OK)
async def readiness_check():
    """
    Readiness probe - checks critical dependencies
    """
    try:
        # Quick DynamoDB check
        dynamodb = boto3.client(
            "dynamodb", region_name=os.getenv("AWS_REGION", "ap-southeast-1")
        )
        table_name = os.getenv("DYNAMODB_TABLE_NAME", "typerush-dev-texts")
        dynamodb.describe_table(TableName=table_name)

        return {
            "status": "ready",
            "service": "text-service",
            "timestamp": datetime.utcnow().isoformat(),
        }
    except Exception as e:
        raise HTTPException(
            status_code=503,
            detail={
                "status": "not-ready",
                "service": "text-service",
                "timestamp": datetime.utcnow().isoformat(),
                "error": str(e),
            },
        )

File: services/record-service/src/modules/health/health.controller.ts
import { Controller, Get } from '@nestjs/common';
import { PrismaService } from '../../prisma/prisma.service';

interface HealthCheckResponse {
  status: 'healthy' | 'unhealthy';
  service: string;
  timestamp: string;
  checks: {
    database?: { status: string; latency?: number; error?: string };
  };
  version: string;
}

@Controller('health')
export class HealthController {
  constructor(private readonly prisma: PrismaService) {}

  /**
   * Comprehensive health check with all dependencies
   * GET /health
   */
  @Get()
  async health(): Promise<HealthCheckResponse> {
    const checks: HealthCheckResponse['checks'] = {};
    let overallStatus: 'healthy' | 'unhealthy' = 'healthy';

    try {
      // Check PostgreSQL database connectivity
      const dbStart = Date.now();
      try {
        await this.prisma.$queryRaw`SELECT 1`;
        checks.database = {
          status: 'healthy',
          latency: Date.now() - dbStart,
        };
      } catch (error) {
        checks.database = {
          status: 'unhealthy',
          error: error instanceof Error ? error.message : 'Unknown error',
        };
        overallStatus = 'unhealthy';
      }

      const response: HealthCheckResponse = {
        status: overallStatus,
        service: 'record-service',
        timestamp: new Date().toISOString(),
        checks,
        version: process.env.APP_VERSION || '1.0.0',
      };

      return response;
    } catch (error) {
      return {
        status: 'unhealthy',
        service: 'record-service',
        timestamp: new Date().toISOString(),
        checks,
        version: process.env.APP_VERSION || '1.0.0',
      };
    }
  }

  /**
   * Simple liveness probe - no dependencies checked
   * GET /health/live
   */
  @Get('live')
  liveness() {
    return {
      status: 'alive',
      service: 'record-service',
      timestamp: new Date().toISOString(),
    };
  }

  /**
   * Readiness probe - checks if ready to serve traffic
   * GET /health/ready
   */
  @Get('ready')
  async readiness() {
    try {
      await this.prisma.$queryRaw`SELECT 1`;
      return {
        status: 'ready',
        service: 'record-service',
        timestamp: new Date().toISOString(),
      };
    } catch (error) {
      throw new Error('Not ready');
    }
  }
}
